### PR TITLE
Refactor PFDL code

### PR DESCRIPTION
**Pull request by maxhoerstr**
- Make code more modular to enable the reuse of the code (e.g.. make more functions and methods single purpose)
- Fix various bugs in the petri net generation
- Renaming
- Format code and add docstrings
- Allow a PFDL program to be passed over to the scheduler as a string (currently, only a file path was allowed)
- Add various quality of live features/fixes
- Allow the free configuration of the "start task" (former productionTask)
- Allow integers as loop limits and negative numbers

### CODE REVIEW

File: tests/test_files/scheduling/task_synchronisation.pfdl [hunk 1]
```diff
@@ -20,7 +20,7 @@ Struct CuttingResult
     sheet_parts: SheetPart[]
 End
 
-Struct TransportOrderStep
+Struct TransportOrderStep_Struct
```
Comment (OliverStolzBO) [inline tests/test_files/scheduling/task_synchronisation.pfdl:23]: Maybe remove the _ in the Struct and Service names here to be consistent with the other test files

File: tests/test_files/scheduling/task_synchronisation.pfdl [hunk 2]
```diff
@@ -50,14 +50,14 @@ Task cuttingTask
 End
 
 Task transport
-    Transport
+    Transport_Service
```
Comment (OliverStolzBO) [inline tests/test_files/scheduling/task_synchronisation.pfdl:53]: See comment above

File: tests/unit_test/test_pfdl_tree_visitor.py [hunk 3]
```diff
@@ -651,51 +671,77 @@ def test_visitVariable_definition(self):
...
+        mock.assert_called_with(array_context, "string")
+
+    def test_initializeArray(self):
+        array_context = PFDLParser.ArrayContext(None)
+        variable_type = "string"
+
+        # without length
+        with patch.object(self.visitor, "visitArray", return_value=-1):
+            array = Array("string")
+            array.length = 10
```
Comment (OliverStolzBO) [inline tests/unit_test/test_pfdl_tree_visitor.py:723]: This lines are meaningless and can be removed

**Review by OliverStolzBO — COMMENTED** [review]
Looks good overall, I only figured out some minor improvements that could be applied to increase the code quality.

File: pfdl_scheduler/petri_net/generator.py [hunk 4]
```diff
@@ -7,6 +7,7 @@
 """Contains the PetriNetGenerator class."""
 
 # standard libraries
+import copy
 from typing import Any, Callable, Dict, List, OrderedDict, Tuple
```
[the remark this answers is not in the document]
Comment (maxhoerstr) [reply pfdl_scheduler/petri_net/generator.py:11]: Copy is used in Line 54 :)

[the remark this answers is not in the document]
Comment (OliverStolzBO) [reply pfdl_scheduler/petri_net/generator.py:11]: Sorry I marked the wrong line, this comment refers to the Tuple import in the line below

File: tests/test_files/scheduling/task_synchronisation.pfdl [hunk 5]
```diff
@@ -20,7 +20,7 @@ Struct CuttingResult
     sheet_parts: SheetPart[]
 End
 
-Struct TransportOrderStep_Struct
+Struct TransportOrderStep
```
Comment (OliverStolzBO) [inline tests/test_files/scheduling/task_synchronisation.pfdl:23]: Using keywords of other plugins will cause problems there, so the Struct needs to have a different name than TransportOrderStep (same for Transport)

**Review by OliverStolzBO — CHANGES_REQUESTED** [review]

**Review by OliverStolzBO — APPROVED** [review]

**Pull request merged**